pkg/client/core/service_monitoring_client.go

```diff
@@ -40,28 +50,206 @@ const (
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart := promtail.New(nil)
+
+	err = s.api.DeletePromtail(api.DeleteHelmReleaseOpts{
+		ID:          opts.ID,
+		ReleaseName: chart.ReleaseName,
+		Namespace:   chart.Namespace,
+	})
+	if err != nil {
+		return err
+	}
+
+	report, err := s.store.RemovePromtail(opts.ID)
+	if err != nil {
+		return err
+	}
+
+	return s.report.ReportRemovePromtail(report)
+}
+
+func (s *monitoringService) CreatePromtail(_ context.Context, opts client.CreatePromtailOpts) (*client.Promtail, error) {
+	err := s.spinner.Start("promtail")
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart, err := s.api.CreatePromtail(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	l := &client.Promtail{
+		ID:    opts.ID,
+		Chart: chart,
+	}
+
+	report, err := s.store.SavePromtail(l)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.report.ReportSavePromtail(l, report)
+	if err != nil {
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (s *monitoringService) DeleteLoki(ctx context.Context, opts client.DeleteLokiOpts) error {
+	err := s.spinner.Start("loki")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	chart := loki.New(nil)
+
+	err = s.manifest.DeleteNativeSecret(ctx, client.DeleteNativeSecretOpts{
+		ID:        opts.ID,
+		Name:      lokiDatasourceSecretName,
+		Namespace: config.DefaultMonitoringNamespace,
+	})
+
+	err = s.api.DeleteLoki(api.DeleteHelmReleaseOpts{
+		ID:          opts.ID,
+		ReleaseName: chart.ReleaseName,
+		Namespace:   chart.Namespace,
+	})
+	if err != nil {
+		return err
+	}
+
+	report, err := s.store.RemoveLoki(opts.ID)
 	if err != nil {
 		return err
 	}
 
+	return s.report.ReportRemoveLoki(report)
+}
+
+func (s *monitoringService) CreateLoki(ctx context.Context, opts client.CreateLokiOpts) (*client.Loki, error) {
+	err := s.spinner.Start("loki")
+	if err != nil {
+		return nil, err
+	}
+
 	defer func() {
 		err = s.spinner.Stop()
 	}()
 
-	// This is too heavy handed, we should only remove the chart
-	// and the secrets manifest. We can, however, wait until Loki
-	// with doing this.
-	err = s.manifest.DeleteNamespace(ctx, api.DeleteNamespaceOpts{
+	chart, err := s.api.CreateLoki(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := yaml.Marshal(lokipkg.NewDatasourceTemplate())
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.manifest.CreateNativeSecret(ctx, client.CreateNativeSecretOpts{
 		ID:        opts.ID,
+		Name:      lokiDatasourceSecretName,
 		Namespace: config.DefaultMonitoringNamespace,
+		Data: map[string]string{
+			"loki-datasource.yaml": string(data),
+		},
+		Labels: map[string]string{
+			"grafana_datasource": "1",
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// The datasources are only loaded during grafana startup, so we need
+	// to cycle grafana to have it pick up the changes
+	for _, replicas := range []int32{0, 1} {
```

Comment:
In loki_reconciler.go, we run CreateLoki every time, which means that grafana is scaled to 0 even though there might be no changes - is this correct? I'm guessing it's not worth the effort to make a better check here, but I'm mentioning in case you know of a super easy way of doing it.

Reply:
I thought `CreateLoki` would only run if the current and expected trees were different, so if the `loki` directories aren't present it will run the check? If this is not the case, this isn't super pretty, I agree. I do believe that there is a Grafana API that can be called with the datasource instead, at least I have seen examples of this. I am not sure that it is worth the overhead given the minimal amount of impact this will have?

Reply:
> I thought CreateLoki would only run if the current and expected trees were different, so if the loki directories aren't present it will run the check?

That's right, I forgot about that.

No need to do anything here for now.